vnext/Microsoft.ReactNative.Cxx/NativeModules.h

```diff
@@ -453,19 +639,103 @@ struct ModuleEventFieldInfo<TFunc<void(TArg)> TModule::*> {
-          module->*field = [eventHandler](TArg arg) noexcept {
-            eventHandler([&](const IJSValueWriter &argWriter) noexcept { WriteValue(argWriter, arg); });
-          };
-        });
+struct ReactModuleBuilder {
+  ReactModuleBuilder(void *module, IReactModuleBuilder const &moduleBuilder) noexcept
+      : m_module{module}, m_moduleBuilder{moduleBuilder} {}
+
+  template <class TModule, int I>
+  void RegisterModule(
+      const wchar_t * /*moduleName*/,
+      const wchar_t *eventEmitterName,
+      winrt::Microsoft::ReactNative::Bridge::ReactMemberId<I>) noexcept {
+    m_moduleBuilder.SetEventEmitterName(eventEmitterName);
+    RegisterMembers<TModule, I>();
+  }
 
-    return false;
+  template <class TClass, int I>
+  auto HasRegisterMember(ReactModuleBuilder &builder, ReactMemberId<I> id)
+      -> decltype(TClass::template RegisterMember<TClass>(builder, id), std::true_type{});
+  template <class TClass>
+  auto HasRegisterMember(...) -> std::false_type;
+
+  template <class TModule, int I>
+  void RegisterMembers() noexcept {
+    if constexpr (decltype(HasRegisterMember<TModule>(*this, ReactMemberId<I + 1>{}))::value) {
+      TModule::template RegisterMember<TModule>(*this, ReactMemberId<I + 1>{});
+      RegisterMembers<TModule, I + 1>();
+    }
+  }
+
+  template <class TClass, class TMethod>
+  void RegisterMethod(TMethod method, wchar_t const *name) noexcept {
+    MethodReturnType returnType;
+    auto methodDelegate = ModuleMethodInfo<TMethod>::GetMethodDelegate(m_module, method, /*out*/ returnType);
+    m_moduleBuilder.AddMethod(name, returnType, methodDelegate);
+  }
+
+  template <class TClass, class TMethod>
+  void RegisterSyncMethod(TMethod method, wchar_t const *name) noexcept {
+    auto syncMethodDelegate = ModuleSyncMethodInfo<TMethod>::GetMethodDelegate(m_module, method);
+    m_moduleBuilder.AddSyncMethod(name, syncMethodDelegate);
   }
+
+  template <class TClass, class TMethod>
+  void RegisterConstMethod(TMethod method, wchar_t const * /*name*/) noexcept {
+    auto constantProvider = ModuleConstantInfo<TMethod>::GetConstantProvider(m_module, method);
+    m_moduleBuilder.AddConstantProvider(constantProvider);
+  }
+
+  template <class TClass, class TField>
+  void RegisterConstant(TField field, wchar_t const *name) noexcept {
+    auto constantProvider = ModuleConstFieldInfo<TField>::GetConstantProvider(m_module, name, field);
+    m_moduleBuilder.AddConstantProvider(constantProvider);
+  }
+
+  template <class TClass, class TField>
+  void RegisterEvent(TField field, wchar_t const *name) noexcept {
+    auto eventHandlerSetter = ModuleEventFieldInfo<TField>::GetEventHandlerSetter(m_module, field);
+    m_moduleBuilder.AddEventHandlerSetter(name, eventHandlerSetter);
+  }
+
+ private:
+  void *m_module;
+  IReactModuleBuilder m_moduleBuilder;
+};
+
+template <class T>
+struct BoxedValue : implements<BoxedValue<T>, IBoxedValue> {
+  BoxedValue() noexcept {}
+
+  int64_t GetPtr() noexcept {
```

Comment:
>int64_t [](start = 2, length = 7)

Is there a way to specify in IDL to use ptr width types (e.g. intptr_t). Using int64_t for pointers can get annoying if you ever need to do assignment on x86 since the compiler will worry about truncation.

Reply:
Agee. But, no, the new IDL 3.0 does not allow use of pointers or integers of pointer size. [https://docs.microsoft.com/en-us/uwp/midl-3/intro](https://docs.microsoft.com/en-us/uwp/midl-3/intro)
It is very similar to C# in that sense.

---
In reply to: [356920177](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356920177) [](ancestors = 356920177)